### PR TITLE
Module #3

### PR DESCRIPTION
**Pull request by evibart**
Added screen with the movie list from JSON and navigation through pages

https://github.com/evibart/TheMovieDBApp/assets/142030257/007c5b08-274f-478f-bb83-2d92d91c9a51

### CODE REVIEW

File: lib/data_classes/movie.dart [hunk 1]
```diff
@@ -1,45 +1,79 @@
...
+  factory Movie.fromJson(Map<String, dynamic> movieJson) {
+    return Movie(
+      title: movieJson["title"],
+      originalTitle: movieJson["original_title"],
+      overview: movieJson["overview"],
+      releaseDate: DateTime.parse(movieJson["release_date"]),
+      voteAverage: movieJson["vote_average"].toDouble(),
+      genres: movieJson["genre_ids"].cast<int>(),
+      poster: imageBaseURL + movieJson["poster_path"],
+      backdrop: imageBaseURL + movieJson["backdrop_path"],
```
Comment (matiasilveiro-gl) [inline lib/data_classes/movie.dart]: Better move the string interpolation of the base URL to custom getters:
`String get posterUrl => imageBaseUrl + poster`

File: lib/movie_widgets/movie_app_bar.dart [hunk 2]
```diff
@@ -17,8 +19,15 @@ class MovieAppBar extends StatelessWidget implements PreferredSizeWidget {
       toolbarHeight: kToolbarHeight,
       title: Text(movieTitle),
       leading: IconButton(
-        onPressed: () {},
-        icon: const Icon(Icons.arrow_back),
+        onPressed: () {
+          Navigator.pushNamed(
+            context,
+            MovieListHome.routeName,
+          );
```
Comment (matiasilveiro-gl) [inline lib/movie_widgets/movie_app_bar.dart]: I'm not so sure, but I think this should be `Navigator.of(context).pop()`, because each time you're pressing this "back" button you're not going "back", but pushing a new screen to the stack. 
Try tapping this button, and then Android's back button and see how your app behaves!

File: lib/movie_widgets/movie_genres.dart [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+    }
+    return genresList;
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    IRepository genresRepository = GenresLocalRepository();
+
+    return FutureBuilder(
+      future: genresRepository.loadData(),
```
Comment (matiasilveiro-gl) [inline lib/movie_widgets/movie_genres.dart]: We can add some Flutter best practices here. Please read this:
https://api.flutter.dev/flutter/widgets/FutureBuilder-class.html

> If the future is created at the same time as the FutureBuilder, then every time the FutureBuilder's parent is rebuilt, the asynchronous task will be restarted. A general guideline is to assume that every build method could get called every frame, and to treat omitted calls as an optimization.


TLDR, you could make this widget Stateful, declare a Future<T> movies at the beginning, init that future in the initState method, and use that future in the FutureBuilder. Like here:
https://stackoverflow.com/a/68966640

File: lib/repositories/genres_local_repository.dart [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+
+import 'package:flutter/services.dart';
+
+import '/repositories_i/repository_i.dart';
+
+class GenresLocalRepository implements IRepository {
+  static const String _jsonGenresPath = 'movie_data/genres.json';
+
+  @override
+  Future<Map<int, String>> loadData() async {
```
Comment (matiasilveiro-gl) [inline lib/repositories/genres_local_repository.dart]: Better create a `Genre` class for this, and return a `List<Genre>`

File: lib/screens/movie_home_page.dart [hunk 5]
```diff
@@ -1,18 +1,21 @@
...
-  MovieHomePage({super.key});
+  const MovieHomePage({super.key});
 
-  final Movie movie = Movie.defaultMovie();
+  static const String routeName = "/movie";
 
   @override
   Widget build(BuildContext context) {
+    final Movie? movieFromNamed =
+        ModalRoute.of(context)?.settings.arguments as Movie?;
```
Comment (matiasilveiro-gl) [inline lib/screens/movie_home_page.dart]: Don't declare this as nullable if you'll force-unwrap it later. Either cast as `Movie!` here or check that if it's null show an error message.
Personally, I'd take the second approach :)

File: lib/screens/movie_list_home.dart [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+  static const double movieBoxSize = 125;
+  static const double titleSize = 20;
+  static const int overviewMaxLines = 3;
+
+  @override
+  Widget build(BuildContext context) {
+    IRepository movieRepository = MovieLocalRepository();
+    return Scaffold(
+      body: FutureBuilder(
+        future: movieRepository.loadData(),
```
Comment (matiasilveiro-gl) [inline lib/screens/movie_list_home.dart]: Same as before

File: lib/data_classes/movie.dart (hunk 1 above)
Comment (evibart) [reply lib/data_classes/movie.dart]: fixed

File: lib/movie_widgets/movie_app_bar.dart (hunk 2 above)
Comment (evibart) [reply lib/movie_widgets/movie_app_bar.dart]: fixed

File: lib/movie_widgets/movie_genres.dart (hunk 3 above)
Comment (evibart) [reply lib/movie_widgets/movie_genres.dart]: fixed

File: lib/repositories/genres_local_repository.dart (hunk 4 above)
Comment (evibart) [reply lib/repositories/genres_local_repository.dart]: fixed

File: lib/screens/movie_home_page.dart (hunk 5 above)
Comment (evibart) [reply lib/screens/movie_home_page.dart]: fixed

File: lib/screens/movie_list_home.dart (hunk 6 above)
Comment (evibart) [reply lib/screens/movie_list_home.dart]: fixed

File: lib/movie_widgets/movie_genres.dart [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+  late Future<List<Genre>> genresList;
+
+  @override
+  initState() {
+    super.initState();
+    genresList = genreRepository.loadData() as Future<List<Genre>>;
+  }
+  @override
+  Widget build(BuildContext context) {
+    return FutureBuilder(
```
Comment (LeonardoMolleker) [inline lib/movie_widgets/movie_genres.dart]: define the type of the future builder

File: lib/movie_widgets/movie_genres.dart [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+      future: genresList,
+      builder: (
+          BuildContext context,
+          snapshot,
```
Comment (LeonardoMolleker) [inline lib/movie_widgets/movie_genres.dart]: same for snapshot

File: lib/movie_widgets/movie_genres.dart [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+      ) {
+        if (snapshot.hasError) {
+          return Center(
+            child: Text(
+              "${snapshot.error}",
+            ),
+          );
+        } else if (snapshot.hasData) {
+          List<String> genresList = widget.linkGenres(snapshot.data as List<Genre>);
```
Comment (LeonardoMolleker) [inline lib/movie_widgets/movie_genres.dart]: definining above the type of the snapshot this cast probably won't necessary

File: lib/screens/movie_home_page.dart [hunk 10]
```diff
@@ -8,14 +8,20 @@ class MovieHomePage extends StatelessWidget {
   const MovieHomePage({super.key});
 
   static const String routeName = "/movie";
+  static const String nullError = "The argument of navigation isn't a movie";
```
Comment (Kevin-Guillen) [inline lib/screens/movie_home_page.dart]: As a user if a get this message i would not understand whats happening

File: lib/screens/movie_list_home.dart [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+  @override
+  initState() {
+    super.initState();
+    movies = movieRepository.loadData() as Future<List<Movie>>;
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    return Scaffold(
+      body: FutureBuilder(
```
Comment (LeonardoMolleker) [inline lib/screens/movie_list_home.dart]: again, define the type

File: lib/screens/movie_list_home.dart [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+        future: movies,
+        builder: (
+          BuildContext context,
+          snapshot,
```
Comment (LeonardoMolleker) [inline lib/screens/movie_list_home.dart]: same as above

File: lib/screens/movie_list_home.dart [hunk 13]
```diff
@@ -0,0 +1,56 @@
...
+        ) {
+          if (snapshot.hasError) {
+            return Center(
+              child: Text(
+                "${snapshot.error}",
+              ),
+            );
+          } else if (snapshot.hasData) {
+            List<Movie> moviesList = snapshot.data as List<Movie>;
```
Comment (LeonardoMolleker) [inline lib/screens/movie_list_home.dart]: check if after defining type the cast is still necessary

File: lib/movie_widgets/movie_genres.dart (hunk 7 above)
Comment (evibart) [reply lib/movie_widgets/movie_genres.dart]: fixed

File: lib/movie_widgets/movie_genres.dart (hunk 8 above)
Comment (evibart) [reply lib/movie_widgets/movie_genres.dart]: fixed

File: lib/movie_widgets/movie_genres.dart (hunk 9 above)
Comment (evibart) [reply lib/movie_widgets/movie_genres.dart]: fixed

File: lib/screens/movie_list_home.dart (hunk 11 above)
Comment (evibart) [reply lib/screens/movie_list_home.dart]: fixed

File: lib/screens/movie_home_page.dart (hunk 10 above)
Comment (evibart) [reply lib/screens/movie_home_page.dart]: fixed

File: lib/screens/movie_list_home.dart (hunk 12 above)
Comment (evibart) [reply lib/screens/movie_list_home.dart]: fixed

File: lib/screens/movie_list_home.dart (hunk 13 above)
Comment (evibart) [reply lib/screens/movie_list_home.dart]: fixed

File: lib/movie_widgets/movie_genres.dart [hunk 14]
```diff
@@ -0,0 +1,80 @@
...
+}
+
+class _MovieGenresState extends State<MovieGenres> {
+  final IRepository genreRepository = GenreLocalRepository();
+  late Future<List<Genre>> genresList;
+
+  @override
+  initState() {
+    super.initState();
+    genresList = genreRepository.loadData() as Future<List<Genre>>;
```
Comment (matiasilveiro-gl) [inline lib/movie_widgets/movie_genres.dart]: I'd move this cast above:
`final IRepository<List<Genre>> genreRepository = GenreLocalRepository();`

File: lib/screens/movie_list_home.dart [hunk 15]
```diff
@@ -0,0 +1,56 @@
...
+}
+
+class _MovieListHomeState extends State<MovieListHome> {
+  final IRepository movieRepository = MovieLocalRepository();
+  late Future<List<Movie>> movies;
+
+  @override
+  initState() {
+    super.initState();
+    movies = movieRepository.loadData() as Future<List<Movie>>;
```
Comment (matiasilveiro-gl) [inline lib/screens/movie_list_home.dart]: Same as above

**Review by matiasilveiro-gl — APPROVED** [review]
LGTM! Minor comments.

**Review by LeonardoMolleker — APPROVED** [review]

File: lib/data_classes/movie.dart [hunk 16]
```diff
@@ -1,45 +1,82 @@
...
         2023,
         4,
         5,
       ),
-      7.8,
-      ["Animation", "Family", "Adventure", "Fantasy", "Comedy"],
-      'movie_images/poster.jpg',
-      'movie_images/backdrop.jpg',
+      voteAverage: 7.8,
+      genres: [16, 10751, 12, 14, 35,],
```
Comment (FranciscoGaraventa) [inline lib/data_classes/movie.dart]: format

File: lib/movie_list_widgets/movie_list.dart [hunk 17]
```diff
@@ -0,0 +1,108 @@
...
+    required this.moviesList,
+  });
+
+  @override
+  Widget build(BuildContext context) {
+    return ListView.builder(
+      itemCount: moviesList.length,
+      itemBuilder: (
+        context,
+        index,
```
Comment (FranciscoGaraventa) [inline lib/movie_list_widgets/movie_list.dart]: As a recommendation add the type of the parameters (BuildContext, int)

File: lib/movie_list_widgets/movie_list.dart [hunk 18]
```diff
@@ -0,0 +1,108 @@
...
+                          child: Text(
+                            moviesList[index].overview,
+                            maxLines: overviewMaxLines,
+                            overflow: TextOverflow.ellipsis,
+                          ),
+                        ),
+                        Padding(
+                          padding: const EdgeInsets.only(
+                            left: Constants.paddingSize,
+                            right: Constants.paddingSize,
```
Comment (FranciscoGaraventa) [inline lib/movie_list_widgets/movie_list.dart]: If you only need to modify left and right you can use `EdgeInsets.symmetric(horizontal: Constants.paddingSize)`

File: lib/movie_widgets/movie_genres.dart [hunk 19]
```diff
@@ -0,0 +1,80 @@
+import 'package:flutter/material.dart';
+import 'package:movie_db_app/repositories_i/repository_i.dart';
```
Comment (FranciscoGaraventa) [inline lib/movie_widgets/movie_genres.dart]: relative import

File: lib/screens/movie_home_page.dart [hunk 20]
```diff
@@ -1,15 +1,24 @@
...
+  static const String nullError = "We had an issue loading the movie";
 
   @override
   Widget build(BuildContext context) {
+    final arg = ModalRoute.of(context)?.settings.arguments;
+    if (arg is! Movie) {
+      return const Text(
+        nullError,
+      );
+    }
```
Comment (FranciscoGaraventa) [inline lib/screens/movie_home_page.dart]: To void this validation here, you can use the body, to show or the error text or the corresponding movie, like this:
`Scaffold(
     appBar: ....,
     body: movie != null ? MovieBody(movie: movie) : Text(errorMsg)
     `

File: lib/screens/movie_list_home.dart [hunk 21]
```diff
@@ -0,0 +1,56 @@
...
+            return Center(
+              child: Text(
+                "${snapshot.error}",
+              ),
+            );
+          } else if (snapshot.hasData) {
+            List<Movie> moviesList = snapshot.data!;
+            return MovieList(
+              moviesList: moviesList,
+            );
```
Comment (FranciscoGaraventa) [inline lib/screens/movie_list_home.dart]: You can use direclty snapshot.data! as a parameter of the MovieList

File: pubspec.yaml [hunk 22]
```diff
@@ -90,3 +90,4 @@ flutter:
   # see https://flutter.dev/custom-fonts/#from-packages
   assets:
     - movie_images/
+    - movie_data/
```
Comment (FranciscoGaraventa) [inline pubspec.yaml]: Can we changed this folder name to mock_data, because you are getting movie.json but also genres.json, not only movie data

File: lib/movie_widgets/movie_genres.dart (hunk 14 above)
Comment (evibart) [reply lib/movie_widgets/movie_genres.dart]: fixed

File: lib/screens/movie_list_home.dart (hunk 15 above)
Comment (evibart) [reply lib/screens/movie_list_home.dart]: fixed

File: lib/data_classes/movie.dart (hunk 16 above)
Comment (evibart) [reply lib/data_classes/movie.dart]: fixed

File: lib/movie_list_widgets/movie_list.dart (hunk 17 above)
Comment (evibart) [reply lib/movie_list_widgets/movie_list.dart]: fixed

File: lib/movie_list_widgets/movie_list.dart (hunk 18 above)
Comment (evibart) [reply lib/movie_list_widgets/movie_list.dart]: fixed

File: lib/movie_widgets/movie_genres.dart (hunk 19 above)
Comment (evibart) [reply lib/movie_widgets/movie_genres.dart]: fixed

File: lib/screens/movie_home_page.dart (hunk 20 above)
Comment (evibart) [reply lib/screens/movie_home_page.dart]: reviewed

File: lib/screens/movie_list_home.dart (hunk 21 above)
Comment (evibart) [reply lib/screens/movie_list_home.dart]: fixed

File: pubspec.yaml (hunk 22 above)
Comment (evibart) [reply pubspec.yaml]: fixed

**Review by Kevin-Guillen — APPROVED** [review]

**Review by FranciscoGaraventa — APPROVED** [review]
LG
